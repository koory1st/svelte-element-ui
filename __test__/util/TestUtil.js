import '@testing-library/jest-dom/extend-expect'

export function testElementAndResult(element) {
  if (!element) {
    throw new Error('Element is empty!')
  }

  expect(element).toMatchSnapshot()
}
