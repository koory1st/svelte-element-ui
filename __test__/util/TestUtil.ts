import '@testing-library/jest-dom/extend-expect'

export function testElementAndResult(element: any) {
  if (!element) {
    throw new Error('Element is empty!')
  }

  expect(element).toMatchSnapshot()
}
