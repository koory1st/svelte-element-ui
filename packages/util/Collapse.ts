/**
 * origin package: svelte-collapse
 */
interface Param {
  open: boolean
  direction?: string
  duration?: number
  easing?: string
}

export default function collapse(node, params: Param) {
  const defaultParams: Param = {
    open: true,
    direction: 'vertical',
    duration: 0.2,
    easing: 'ease',
  }

  params = Object.assign(defaultParams, params)

  const noop = (callback: FrameRequestCallback) => {}
  let transitionEndResolve = noop
  let transitionEndReject = noop

  const listener = node.addEventListener('transitionend', () => {
    transitionEndResolve(null)
    transitionEndResolve = noop
    transitionEndReject = noop
  })

  // convenience functions
  async function asyncTransitionEnd() {
    return new Promise((resolve, reject) => {
      transitionEndResolve = resolve
      transitionEndReject = reject
    })
  }

  async function nextFrame() {
    return new Promise(requestAnimationFrame)
  }

  function transition() {
    const st = params.direction === 'vertical' ? 'height' : 'width'
    return `${st} ${params.duration}s ${params.easing}`
  }

  // set initial styles
  node.style.overflow = 'hidden'
  node.style.transition = transition()
  if (params.direction === 'vertical') {
    node.style.height = params.open ? 'auto' : '0px'
  } else {
    node.style.width = params.open ? 'auto' : '0px'
  }

  async function enter() {
    // height is already in pixels
    // start the transition
    if (params.direction === 'vertical') {
      node.style.height = node.scrollHeight + 'px'
    } else {
      node.style.width = node.scrollWidth + 'px'
    }

    // wait for transition to end,
    // then switch back to height auto
    try {
      await asyncTransitionEnd()
      if (params.direction === 'vertical') {
        node.style.height = 'auto'
      } else {
        node.style.width = 'auto'
      }
    } catch (err) {
      // interrupted by a leave transition
    }
  }

  async function leave() {
    if (params.direction === 'vertical') {
      if (node.style.height === 'auto') {
        // temporarily turn transitions off
        node.style.transition = 'none'
        await nextFrame()

        // set height to pixels, and turn transition back on
        node.style.height = node.scrollHeight + 'px'
        node.style.transition = transition()
        await nextFrame()

        // start the transition
        node.style.height = '0px'
      } else {
        // we are interrupting an enter transition
        transitionEndReject(null)
        node.style.height = '0px'
      }
    } else {
      if (node.style.width === 'auto') {
        // temporarily turn transitions off
        node.style.transition = 'none'
        await nextFrame()

        // set height to pixels, and turn transition back on
        node.style.width = node.scrollWidth + 'px'
        node.style.transition = transition()
        await nextFrame()

        // start the transition
        node.style.width = '0px'
      } else {
        // we are interrupting an enter transition
        transitionEndReject(null)
        node.style.width = '0px'
      }
    }
  }

  function update(newParams) {
    params = Object.assign(params, newParams)
    params.open ? enter() : leave()
  }

  function destroy() {
    node.removeEventListener('transitionend', listener)
  }

  return { update, destroy }
}
