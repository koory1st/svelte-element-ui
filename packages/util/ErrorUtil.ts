export class SeuPropValidateError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'SeuPropValidateError'
  }
}
