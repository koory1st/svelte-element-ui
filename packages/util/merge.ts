export default function (target, ...rest) {
  for (let i = 1, j = rest.length; i < j; i++) {
    let source = rest[i] || {}
    for (let prop in source) {
      if (source.hasOwnProperty(prop)) {
        let value = source[prop]
        if (value !== undefined) {
          target[prop] = value
        }
      }
    }
  }

  return target
}
