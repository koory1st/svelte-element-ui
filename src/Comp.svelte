<script type="ts">
  export let name: String

  let buttonText = 'Button'

  function handleClick() {
    buttonText = 'Button Clicked'
  }
</script>

<h1>Hello {name}!</h1>

<button on:click={handleClick}>{buttonText}</button>
