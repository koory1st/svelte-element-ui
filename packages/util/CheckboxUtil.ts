import { SeuPropValidateError } from './ErrorUtil'
/**
 * getDisabled
 *
 * only used in group mode
 *
 * @param disabledProp
 * @param groupFlg
 * @param group
 * @param max
 * @param min
 * @param innerChecked
 */
export function getDisabled(
  disabledProp: boolean,
  groupFlg: boolean,
  group: Array<string | number | boolean>,
  maxInput: string | number | null,
  minInput: string | number | null,
  innerChecked: boolean,
): boolean {
  if (!groupFlg) {
    return disabledProp
  }

  const max = Number(maxInput)
  const min = Number(minInput)

  if (max && group.length >= max && !innerChecked) {
    return true
  }

  if (min && group.length <= min && innerChecked) {
    return true
  }
  return disabledProp
}

export function getInnerCheckedValue(
  isGroup: boolean,
  checkedValue: string | number | boolean | null | undefined,
  label: string | number | boolean | null | undefined,
): boolean | string | number | null {
  // not in group
  if (!isGroup) {
    if (checkedValue !== null && checkedValue !== undefined) {
      return checkedValue
    }
    return null
  }

  // in group
  if (checkedValue !== null && checkedValue !== undefined) {
    return checkedValue
  }

  if (label !== null && label !== undefined) {
    return label
  }
  return null
}

export function getValueByInnerChecked(
  innerChecked: boolean,
  innerCheckedValue: boolean | string | number | null,
  uncheckedValue: string | number | boolean | null | undefined,
): boolean | string | number {
  if (innerChecked) {
    if (innerCheckedValue === null) {
      return true
    }
    return innerCheckedValue
  } else {
    if (uncheckedValue === null || uncheckedValue === undefined) {
      return false
    }
    return uncheckedValue
  }
}

export function validateCheckedValue(
  checkedValue: string | number | boolean | null | undefined,
  uncheckedValue: string | number | boolean | null | undefined,
) {
  if (checkedValue === null) {
    return
  }
  if (checkedValue === undefined) {
    return
  }
  if (uncheckedValue === null) {
    return
  }
  if (uncheckedValue === undefined) {
    return
  }

  if (checkedValue === uncheckedValue) {
    throw new SeuPropValidateError('checkedValue is the same with uncheckedValue')
  }
}

export function getGroupByInnerChecked(
  isGroup: boolean,
  innerChecked: boolean,
  group: Array<string | number | boolean>,
  innerCheckedValue: boolean | string | number,
): Array<string | number | boolean> {
  if (!isGroup) {
    return group
  }

  if (innerCheckedValue === null) {
    return group
  }

  const groupSet = new Set(group)
  if (innerChecked) {
    groupSet.add(innerCheckedValue)
    return Array.from(groupSet)
  }

  groupSet.delete(innerCheckedValue)
  return Array.from(groupSet)
}

export function getInnerCheckedByValue(
  isGroup: boolean,
  value: boolean | string | number,
  group: Array<string | number | boolean>,
  innerCheckedValue: boolean | string | number | null,
): boolean {
  if (!isGroup) {
    if (innerCheckedValue === null) {
      return Boolean(value)
    } else {
      return value === innerCheckedValue
    }
  } else {
    const groupSet = new Set(group)
    return groupSet.has(innerCheckedValue)
  }
}
