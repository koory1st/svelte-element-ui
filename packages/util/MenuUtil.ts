import { SeuPropValidateError } from './ErrorUtil'
import { mixColor } from './ColorUtil'

export class MenuProps {
  mode?: string = 'vertical'
  collapse?: boolean = false
  backgroundColor?: string = '#ffffff'
  textColor?: string = '#303133'
  activeTextColor?: string = '#409EFF'
  defaultActive?: string
  defaultOpeneds?: Array<string>
  uniqueOpened?: boolean = false
  menuTrigger?: string = 'hover'
  router?: boolean = false
  collapseTransition?: boolean = true
}

export class Menu {
  public constructor(props: MenuProps) {
    this.props = props
  }
  props: MenuProps
  private items: Array<MenuItem> = []
  activeItem: MenuItem
  openedMenusIndex: Array<string> = []

  get hoverBackground() {
    return this.props.backgroundColor ? mixColor(this.props.backgroundColor, 0.2) : null
  }

  public setActive(item: MenuItem): Menu {
    this.activeItem = item
    return this
  }

  public addItem(item: MenuItem): void {
    this.isDuplicate(item)
    this.items.push(item)

    // set default active
    this.setDefaultActive(item)
  }

  private setDefaultActive(item: MenuItem): void {
    if (!this.props.defaultActive) {
      return
    }

    // already set active, return
    if (this.activeItem) {
      return
    }

    if (this.props.defaultActive === item.props.index) {
      this.activeItem = item
    }
  }
  /**
   * check item is the active one
   * @param item
   */
  public isActiveItem(item: MenuItem): boolean {
    if (!item || !this.activeItem) {
      return false
    }
    return this.activeItem === item
  }

  private isDuplicate(item: MenuItem): void {
    if (!item.props.index) {
      return
    }
    if (this.items.map(i => i.props.index).includes(item.props.index)) {
      throw new SeuPropValidateError('menu item index is duplicated: ' + item.props.index)
    }
  }
}

export class MenuItemProps {
  index?: string
  route?: object
  disabled?: boolean = false
}

export class MenuItem {
  props: MenuItemProps
  private parent: Menu

  public constructor(props: MenuItemProps, activeFlg: boolean = false) {
    this.props = props
  }

  public setParent(parent: Menu) {
    this.parent = parent
  }
}

// export enum EnumMode {
//   horizontal = 'horizontal',
//   vertical = 'vertical',
// }

// export enum EnumMenuTrigger {
//   hover = 'hover',
//   click = 'click',
// }
