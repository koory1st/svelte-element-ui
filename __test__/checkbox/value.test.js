import '@testing-library/jest-dom/extend-expect'
import { render, fireEvent } from '@testing-library/svelte'
import { testElementAndResult } from '../util/TestUtil'
import Comp from './value.svelte'

test('toggle the value prop by change prop', async () => {
  const { getByText } = render(Comp)
  const checkboxEle = getByText('test component').querySelector('label')
  testElementAndResult(checkboxEle)

  // toggle the value by change value
  const button = getByText('button02')
  await fireEvent.click(button)
  testElementAndResult(checkboxEle)

  // toggle the value by change value
  await fireEvent.click(button)
  testElementAndResult(checkboxEle)
})
