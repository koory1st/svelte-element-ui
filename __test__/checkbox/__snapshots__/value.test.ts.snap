// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toggle the value prop by change prop 1`] = `
<label
  class="seu-checkbox"
  role="checkbox"
>
  <input
    aria-hidden="false"
    class="seu-checkbox__original"
    type="checkbox"
  />
   
  <span
    class="seu-checkbox__input"
  >
    <span
      class="seu-checkbox__inner"
    />
  </span>
   
  <span
    class="seu-checkbox__label"
  >
    label2
  </span>
</label>
`;

exports[`toggle the value prop by change prop 2`] = `
<label
  class="seu-checkbox is-checked"
  role="checkbox"
>
  <input
    aria-hidden="false"
    class="seu-checkbox__original"
    type="checkbox"
  />
   
  <span
    class="seu-checkbox__input is-checked"
  >
    <span
      class="seu-checkbox__inner"
    />
  </span>
   
  <span
    class="seu-checkbox__label"
  >
    label2
  </span>
</label>
`;

exports[`toggle the value prop by change prop 3`] = `
<label
  class="seu-checkbox"
  role="checkbox"
>
  <input
    aria-hidden="false"
    class="seu-checkbox__original"
    type="checkbox"
  />
   
  <span
    class="seu-checkbox__input"
  >
    <span
      class="seu-checkbox__inner"
    />
  </span>
   
  <span
    class="seu-checkbox__label"
  >
    label2
  </span>
</label>
`;
